pkg/endpoint/bpf.go

```diff
@@ -242,31 +251,152 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
 	return nil, finalizeList.Finalize, revertStack.Revert
 }
 
+func (e *Endpoint) addVisibilityRedirects(ingress bool, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
+	var (
+		visPolicy               policy.DirectionalVisibilityPolicy
+		direction               trafficdirection.TrafficDirection
+		policyEnabled           bool
+		finalizeList            revert.FinalizeList
+		revertStack             revert.RevertStack
+		updatedStats            []*models.ProxyStatistics
+		insertedDesiredMapState = make(map[policy.Key]struct{})
+	)
+
+	if e.visibilityPolicy == nil {
+		return nil, finalizeList.Finalize, revertStack.Revert
+	}
+
+	if ingress {
+		visPolicy = e.visibilityPolicy.Ingress
+		direction = trafficdirection.Ingress
+		policyEnabled = e.desiredPolicy.IngressPolicyEnabled
+	} else {
+		visPolicy = e.visibilityPolicy.Egress
+		direction = trafficdirection.Egress
+		policyEnabled = e.desiredPolicy.EgressPolicyEnabled
+	}
+
+	if !policyEnabled {
+		for _, visMeta := range visPolicy {
+			// Create a redirect for every entry in the visibility policy.
+			if !e.hasSidecarProxy || visMeta.Parser != policy.ParserTypeHTTP {
+				var (
```

Comment:
A lot of this code seems similar to the code present in `addNewRedirectsFromMap`. Can't we refactor it?

Reply:
The first iteration of this code handles the case in which policy is not-enabled. Future work will work on adding redirects based off of the state of the L4Policy which is generated for the Endpoint, which I believe will cause divergence between `addNewVisibilityRedirects` and `addNewRedirectsFromMap`. So, I would prefer keeping them separate for now.